File: packages/agents/src/engine/react-executor.ts
import type {
  AgentDefinitionConfig,
  AgentContext,
  RunState,
  AgentDecision,
  ToolCallRequest,
  ChatMessage,
} from "@testpilot/types";
import { createLogger } from "@testpilot/shared";
import { toolRegistry } from "../tools/registry.js";
import { toolExecutor } from "../tools/tool-executor.js";
import { memoryStore } from "../memory/memory-store.js";
import { telemetry } from "../telemetry/telemetry.js";
import { Groq } from "groq-sdk";

const logger = createLogger("react-executor");

/**
 * Single Reusable ReAct Execution Engine.
 * Replaces duplicated ReAct loops across individual agents.
 * Runs agent configs against Groq LLM with dynamic tool calling and Zod validation.
 */
export class ReActExecutor {
  private groq: Groq;

  constructor() {
    this.groq = new Groq({ apiKey: process.env["GROQ_API_KEY"] || "dummy-key-for-types" });
  }

  /** Execute an agent definition using the central ReAct engine loop */
  public async execute<TOutput = unknown>(
    agentConfig: AgentDefinitionConfig<TOutput>,
    context: AgentContext,
    state: RunState,
    maxSteps = 5
  ): Promise<AgentDecision<TOutput>> {
    const startTime = Date.now();
    logger.info({ agentType: agentConfig.type, runId: state.runId }, "ReActExecutor starting agent execution");

    // Gather allowed tool definitions
    const allowedTools = toolRegistry.getSubset(agentConfig.allowedTools);
    const toolDeclarations = allowedTools.map((t) => ({
      type: "function" as const,
      function: {
        name: t.name,
        description: t.description,
        parameters: t.parameters,
      },
    }));

    // Build chat conversation history with state & memory context
    const runMemories = memoryStore.getMemories(state.runId);
    const messages: ChatMessage[] = [
      { role: "system", content: `${agentConfig.systemPrompt}\nGoal: ${agentConfig.goal}` },
      {
        role: "user",
        content: `Current Run State:\n${JSON.stringify({
          projectId: state.projectId,
          runId: state.runId,
          completedSteps: state.completedSteps,
          memories: runMemories.slice(-5),
        })}`,
      },
    ];

    let currentStep = 0;
    while (currentStep < maxSteps) {
      currentStep++;
      const llmStart = Date.now();

      try {
        const completion = await this.groq.chat.completions.create({
          model: "llama-3.3-70b-versatile",
          messages: messages as any,
          tools: toolDeclarations.length > 0 ? toolDeclarations : undefined,
          temperature: 0.1,
        });

        const reasoningMs = Date.now() - llmStart;
        const tokensUsed = (completion.usage?.prompt_tokens || 0) + (completion.usage?.completion_tokens || 0);
        telemetry.record(state.runId, "reasoning_latency", reasoningMs, agentConfig.type);
        telemetry.record(state.runId, "tokens_used", tokensUsed, agentConfig.type);

        const choice = completion.choices[0]?.message;
        if (!choice) {
          throw new Error("No response message returned from Groq API");
        }

        // Check if LLM requested tool calls
        if (choice.tool_calls && choice.tool_calls.length > 0) {
          const rawCall = choice.tool_calls[0];
          if (!rawCall || !rawCall.function) continue;

          const toolCall: ToolCallRequest = {
            id: rawCall.id || crypto.randomUUID(),
            name: rawCall.function.name,
            arguments: JSON.parse(rawCall.function.arguments || "{}"),
          };

          const toolStart = Date.now();
          const toolResult = await toolExecutor.execute(toolCall, context);
          telemetry.record(state.runId, "tool_latency", Date.now() - toolStart, agentConfig.type, toolCall.name);

          // Store tool execution in MemoryStore & RunState observations
          memoryStore.addMemory(state.runId, toolCall.name, "attempted_fix", toolResult);
          state.observations.push({
            id: crypto.randomUUID(),
            timestamp: new Date(),
            type: "tool_result",
            source: agentConfig.type,
            data: toolResult,
          });

          // If tool completed agent goal, return decision
          return {
            reasoning: choice.content || `Executed tool '${toolCall.name}'`,
            action: toolCall,
            confidence: 0.95,
            completed: true,
            output: toolResult.output as TOutput,
          };
        }

        // Standard completion without tool calls
        return {
          reasoning: choice.content || "Completed agent reasoning task",
          action: null,
          confidence: 0.90,
          completed: true,
          output: { text: choice.content } as unknown as TOutput,
        };
      } catch (err) {
        logger.error({ agentType: agentConfig.type, err }, "ReAct step execution error");
        telemetry.record(state.runId, "failure", 1, agentConfig.type);
        return {
          reasoning: `Execution failed: ${(err as Error).message}`,
          action: null,
          confidence: 0.0,
          completed: false,
        };
      }
    }

    return {
      reasoning: "Reached maximum ReAct step limit without explicit completion",
      action: null,
      confidence: 0.5,
      completed: false,
    };
  }
}

export const reactExecutor = new ReActExecutor();

File: apps/backend/app/agents/react_executor.py
import time
import json
import logging
from typing import Dict, Any, Optional
from groq import Groq
from app.config import settings
from app.models import AgentDefinitionConfig, AgentDecision, RunState, ToolCallRequest
from app.tools.registry import tool_registry
from app.tools.executor import tool_executor
from app.core.telemetry import telemetry
from app.core.memory_store import memory_store

logger = logging.getLogger("react-executor")

class ReActExecutor:
    def __init__(self):
        self.client = Groq(api_key=settings.groq_api_key)

    async def execute(
        self,
        agent_config: AgentDefinitionConfig,
        context: Dict[str, Any],
        state: RunState,
        max_steps: int = 5
    ) -> AgentDecision:
        logger.info(f"ReActExecutor starting execution for agent {agent_config.name}")

        allowed_tools = tool_registry.get_subset(agent_config.allowed_tools)
        tool_declarations = [
            {
                "type": "function",
                "function": {
                    "name": t.name,
                    "description": t.description,
                    "parameters": t.parameters
                }
            }
            for t in allowed_tools
        ]

        messages = [
            {"role": "system", "content": f"{agent_config.system_prompt}\nGoal: {agent_config.description}"},
            {"role": "user", "content": f"Run State: {state.model_dump_json()}"}
        ]

        step = 0
        while step < max_steps:
            step += 1
            start_time = time.time()

            try:
                kwargs = {
                    "model": "llama-3.3-70b-versatile",
                    "messages": messages,
                    "temperature": 0.1
                }
                if tool_declarations:
                    kwargs["tools"] = tool_declarations

                completion = self.client.chat.completions.create(**kwargs)
                latency = time.time() - start_time
                telemetry.record(state.run_id, "reasoning_latency", latency, agent_config.type)

                choice = completion.choices[0].message
                if choice.tool_calls:
                    raw_call = choice.tool_calls[0]
                    tool_call = ToolCallRequest(
                        id=raw_call.id or "call_1",
                        name=raw_call.function.name,
                        arguments=json.loads(raw_call.function.arguments or "{}")
                    )

                    tool_start = time.time()
                    tool_result = await tool_executor.execute(tool_call, context)
                    telemetry.record(state.run_id, "tool_latency", time.time() - tool_start, agent_config.type, tool_call.name)

                    memory_store.add_memory(state.run_id, tool_call.name, "attempted_fix", tool_result.output)
                    return AgentDecision(
                        reasoning=choice.content or f"Executed tool {tool_call.name}",
                        action=tool_call,
                        confidence=0.95,
                        completed=True,
                        output=tool_result.output
                    )

                return AgentDecision(
                    reasoning=choice.content or "Completed reasoning step",
                    action=None,
                    confidence=0.90,
                    completed=True,
                    output={"text": choice.content}
                )

            except Exception as e:
                logger.error(f"ReAct step error: {e}")
                telemetry.record(state.run_id, "failure", 1.0, agent_config.type)
                return AgentDecision(
                    reasoning=f"Execution error: {str(e)}",
                    action=None,
                    confidence=0.0,
                    completed=False
                )

        return AgentDecision(
            reasoning="Reached max steps without explicit completion",
            action=None,
            confidence=0.5,
            completed=False
        )

react_executor = ReActExecutor()
